Update Effective Dart Design Guidelines to Reflect Dart 3.0 Class Modifiers

MiniPiku opened the pull request:
This PR updates the Effective Dart Design Guidelines to align with Dart 3.0's introduction of class modifiers. The changes include:

For API Maintainers:

Removed the recommendation to document whether a class supports being extended or used as an interface.

Added guidance to use class modifiers (e.g., final, sealed, interface, mixin) to explicitly define a class's capabilities.

Emphasized that class modifiers should replace manual documentation for enforcing restrictions on extending or implementing classes.

For Users:

Removed guidelines advising users to avoid extending or implementing classes not meant for it, as these restrictions are now enforced by class modifiers in Dart 3.0.

This change ensures the guidelines are up-to-date with Dart's latest features and best practices.

Fixes: #6437

antfitch commented:
/gcbrun

antfitch commented on src/content/effective-dart/design.md:
```diff
@@ -519,10 +519,24 @@ class in a single library.
-classes. But Dart does not require all code to be defined inside a
-class—you can define top-level variables, constants, and functions like
-you can in a procedural or functional language.
+**Note:** As of Dart 3.0, class modifiers (e.g., `final`, `base`, `interface`) 
```
We should only include very basic top-level information under this high-level heading. Let's revert this to what was already there, but I would advise adding one bullet with a sentence or two that lets people know that classes can be restricted.

 I believe you've already captured these other details in subsections. If something is here but not in one of those sections, let's move the content to that section.

antfitch commented on src/content/effective-dart/design.md:
```diff
@@ -621,41 +635,23 @@ doesn't do that, it's best to assume you should *not* extend the class.
-intend, and they may break your code without realizing it.
 
-### DO document if your class supports being used as an interface
+### DO use class modifiers to control if your class can be used as an interface
```
```suggestion
### DO use class modifiers to control if your class can be an interface
```

I think we can cut a few words from this without removing meaning.

antfitch commented on src/content/effective-dart/design.md:
```diff
@@ -621,41 +635,23 @@ doesn't do that, it's best to assume you should *not* extend the class.
-to implement, then changing those classes becomes very difficult. That
-difficulty in turn means the libraries you rely on are slower to grow and adapt
-to new needs.
+Since Dart 3.0, class modifiers like `final`, `base`, or `sealed` 
```
```suggestion
Class modifiers like `final`, `base`, or `sealed` 
```

Dart 3 has been out for a while. I think we can avoid mentioning it. @johnpryan, your thoughts?

johnpryan replied:
I agree - we can leave this out here.

antfitch commented on src/content/effective-dart/design.md:
```diff
@@ -621,41 +635,23 @@ doesn't do that, it's best to assume you should *not* extend the class.
 
-If your class can be used as an interface, mention that in the class's doc
-comment.
+With Dart 3.0’s class modifiers, you can restrict implementation using
```
Can you have this section mirror the last one as closely as possible? It's okay if they sound very similar. I like the wording you used in the previous heading for extending.

antfitch commented on examples/language/lib/patterns/json.dart:
I see several changes in here related to JSON data. I suspect these are not part of the class modifiers PR. Can you update your PR?

MiniPiku replied:
The json changes are of another PR 
but this PR doesn't contain any of those changes 
this one is created on a separate branch 
idk why its showing like that :/

MiniPiku replied:
the conflict warning has been taken care of :)
a false indentation in 'src/content/language/patterns.md' has been creating the problem
thank you

antfitch reviewed (commented):
Looks great! A few notes added plus it looks like you need to decouple some JSON changes.

johnpryan commented on src/content/effective-dart/design.md:
```diff
@@ -519,10 +519,9 @@ class in a single library.
 
-Dart is a "pure" object-oriented language in that all objects are instances of
-classes. But Dart does not require all code to be defined inside a
+Classes. But Dart does not require all code to be defined inside a
```
The word "class" shouldn't be capitalized here.

```suggestion
classes. But Dart does not require all code to be defined inside a
```

antfitch commented on src/content/effective-dart/design.md:
```diff
@@ -637,20 +622,20 @@ Otherwise, later changes to it may break your code.
 
 
-### DO use class modifiers to control if your class can be used as an interface
+### ### DO use class modifiers to control if your class can be an interface
```
Delete extra ###

antfitch commented on src/content/effective-dart/design.md:
```diff
@@ -637,20 +622,20 @@ Otherwise, later changes to it may break your code.
-`final`, `base`, or `interface`. For example, `final class A {}` 
-prevents implementation, while `interface class C {}` allows it explicitly. 
-Use these modifiers to enforce your design intent instead of relying solely 
+you can also restrict implementation using
```
You can restrict implementation using

johnpryan commented on src/content/effective-dart/design.md:
```diff
@@ -519,10 +519,9 @@ class in a single library.
+Classes. But Dart does not require all code to be defined inside a
 class—you can define top-level variables, constants, and functions like
-you can in a procedural or functional language.
+you can in a procedural or functional language
```
Missing period:

```suggestion
you can in a procedural or functional language.
```

antfitch commented on src/content/effective-dart/design.md:
```diff
@@ -519,24 +519,9 @@ class in a single library.
-- For example, `final class A {}` prevents implementation, while `interface class C {}` allows it explicitly.
-- Use these modifiers to enforce your design intent instead of relying solely on documentation.
-
+Classes. But Dart does not require all code to be defined inside a
```
Original text is still missing.

antfitch commented on examples/language/lib/patterns/json.dart~:
This appears to be an artifact from another PR. Maybe the answer is to start a new PR with your changes for Effective Dart Design? Otherwise, keep working on getting this file (and patterns.md and patterns.md~) removed from the PR.

johnpryan commented on src/content/effective-dart/design.md:
```diff
@@ -621,41 +620,23 @@ doesn't do that, it's best to assume you should *not* extend the class.
-class, state that. Suffix the class name with `Base`, or mention it in the
-class's doc comment.
+Class modifiers like `final`, `base`, or `sealed` 
+allow you to enforce whether a class can be extended directly in code. 
```
```suggestion
enforce whether a class can be extended. 
```

johnpryan commented on src/content/effective-dart/design.md:
```diff
@@ -621,41 +620,23 @@ doesn't do that, it's best to assume you should *not* extend the class.
+allow you to enforce whether a class can be extended directly in code. 
+For example, use `final class A {}` to prevent extension, 
+or `base class B {}` to allow extension only within the same library. 
+Rely on these modifiers rather than documentation to communicate and 
+enforce your intent.
```
```suggestion
Use these modifiers to communicate your intent, rather than relying on documentation.
```

johnpryan commented on src/content/effective-dart/design.md:
```diff
@@ -621,41 +620,23 @@ doesn't do that, it's best to assume you should *not* extend the class.
 
 
-### AVOID implementing a class that isn't intended to be an interface
+### ### DO use class modifiers to control if your class can be an interface
```
Your proposed rule is from the perspective of someone writing the library, but the original version was intended for those who are using a library. This section is talking about classs that aren't intended to be an interface, and while it would be nice if all such classes were using class modifiers to enforce these rules (as recommended above), it's still possible that a developer could run into these problems so I would vote to keep this section as-is.

johnpryan commented on src/content/language/patterns.md~:
```diff
@@ -0,0 +1,447 @@
+---
+title: Patterns
```
Did you intend to make a copy of this file with a `~` at the end? Could you double check that your changes to this file are showing up correctly in the diff?

MiniPiku commented:
@antfitch @johnpryan kindly take a look at the changes made 
https://github.com/dart-lang/site-www/commit/41f58b3542dfc18ec725bc23512d83db1b99a7de
thank you

johnpryan commented on examples/language/lib/patterns/json.dart~:
```diff
@@ -0,0 +1,35 @@
+    }
+    // #enddocregion json-3
+  }
+}
```
Did you intend to include this file?

MiniPiku replied:
nope i don't know why this json.dart~ file is bring mentioned back to back
checked my git status it not showing up here :/

johnpryan replied:
No problem, I can delete it.

MiniPiku commented on src/content/effective-dart/design.md:
```diff
@@ -621,42 +620,22 @@ doesn't do that, it's best to assume you should *not* extend the class.
+Use these modifiers to communicate your intent, rather than relying on documentation.
 
 
-### AVOID implementing a class that isn't intended to be an interface
```
that was the old heading ig 
it was updated to 

 _### DO use class modifiers to control if your class can be an interface_

MiniPiku replied:
just checked my code base 
i guess this was mentioned to be removed and '### DO use class modifiers to control if your class can be an interface'
is suppose to wrap the content of it too

if u need i can add '### AVOID implementing a class that isn't intended to be an interface' and its following contents back in along with '### DO use class modifiers to control if your class can be an interface'

johnpryan replied:
Implicit interfaces are still supported in Dart, so I think this section is still useful. I think you're changing the original meaning a bit more than is necessary. For example:

> To give the authors of the classes you use more leeway, avoid implementing
implicit interfaces except for classes that are clearly intended to be
implemented.

This is still true in Dart 3, it's possible that developers will try to implement a class that wasn't intended to be implemented (and is not marked with a class modifier like `final` ).

MiniPiku replied:
thanks for your input 
lemme rephrase the para a bit that aligns your POV

johnpryan replied:
I think I need to be more clear what I'm asking, thanks for your patience. You are making changes to three rules:

- (Updated) DO use class modifiers to control if your class can be extended
- (Deleted) AVOID implementing a class that isn't intended to be an interface
- (Updated) DO use class modifiers to control if your class can be an interface

I wasn't asking you to rephrase #3, so you can change that back to match #1. I'd prefer to leave this second section as-is for now, and discuss removing it in a separate issue, since it still applies IMO. Thanks!

MiniPiku replied:
made the suggested changes 
kindly take a look

johnpryan commented:
This is looking good, could you take a look and double check that there aren't extra changes committed that are unrelated? I see some changes to .json files that don't seem to belong here.

MiniPiku commented:
yeah i removed the json file that was unnecessarily getting committed to this PR i dont know why
i removed it from my side so you wont have to delete it manually from your side

antfitch commented on src/content/effective-dart/design.md:
```diff
@@ -519,7 +519,6 @@ class in a single library.
 
 ## Classes and mixins
 
-Dart is a "pure" object-oriented language in that all objects are instances of
```
Re-add this.

antfitch commented on src/content/effective-dart/_toc.md:
```diff
@@ -199,10 +199,9 @@ the project:
+* <a href='/effective-dart/design#do-use-class-modifiers-to-control-if-your-class-can-be-extended'>DO use class modifiers to control if your class can be extended</a>
 * <a href='/effective-dart/design#avoid-implementing-a-class-that-isnt-intended-to-be-an-interface'>AVOID implementing a class that isn't intended to be an interface.</a>
-* <a href='/effective-dart/design#do-document-if-your-class-supports-being-used-as-an-interface'>DO document if your class supports being used as an interface.</a>
+* <a href='/effective-dart/design#do-use-class-modifiers-to-control-if-your-class-can-be-used-as-an-interface'>DO use class modifiers to control if your class can be used as an interface.</a>
```
fix link (broke with heading change)

MiniPiku replied:
fixed those but its still giving the 
"anchor link broken" error :/

antfitch reviewed (commented):
Very close now! Just a few tiny changes and we are done. Great job!

MiniPiku commented:
@antfitch kindly take a look at the changes i hope this will work

eernstg commented on src/content/effective-dart/design.md:
```diff
@@ -652,11 +653,13 @@ implicit interfaces except for classes that are clearly intended to be
-comment.
+### DO use class modifiers to control if your class can be an interface
 
+When designing a library, use class modifiers like `final`, `base`, or `interface` to enforce intended
```
When it comes to `implements` clauses, `interface` has no effect, but `sealed` does:
```suggestion
When designing a library, use class modifiers like `final`, `base`, or `sealed` to enforce intended
```

eernstg commented on src/content/effective-dart/design.md:
```diff
@@ -652,11 +653,13 @@ implicit interfaces except for classes that are clearly intended to be
+### DO use class modifiers to control if your class can be an interface
 
+When designing a library, use class modifiers like `final`, `base`, or `interface` to enforce intended
+usage. For example, `final class C {}` prevents implementation, while `interface class D {}`
```
```suggestion
usage. For example, use `final class C {}` or `base class D {}` to prevent
implementation outside the current library.
```

eernstg commented on src/content/effective-dart/design.md:
```diff
@@ -652,11 +653,13 @@ implicit interfaces except for classes that are clearly intended to be
 
+When designing a library, use class modifiers like `final`, `base`, or `interface` to enforce intended
+usage. For example, `final class C {}` prevents implementation, while `interface class D {}`
+explicitly allows it. While it's ideal for all libraries to use these modifiers to enforce design intent,
```
```suggestion
While it's ideal for all libraries to use these modifiers to enforce design intent,
```

eernstg reviewed (commented):
I added a couple of drive-by comments to this PR because I noticed that not all of the class modifiers that were mentioned are relevant to the enclosing text. (For example, `base` is not relevant when the topic is how to control extension. In contrast, the main effect and purpose of using `interface` is precisely to control extension).

MiniPiku commented:
@eernstg kindly take a look as the changes made

eernstg commented on src/content/effective-dart/design.md:
```diff
@@ -621,12 +621,13 @@ doesn't do that, it's best to assume you should *not* extend the class.
 
+Class modifiers like `final`, `interface`, or `sealed`
+restrict how a class can be extended.
+For example, use `final class A {}` or `base class B {}` to prevent 
```
It can't be `base class B {}` because that does not prevent extension outside the current library. It's `interface` that prevents `extends` (but allows `implements`).

MiniPiku replied:
changed `base class B {}` to 'interface class B{}'

eernstg reviewed (commented):
Commented on just one place that needs to say `interface` rather than `base` (`base` doesn't restrict `extends`, it restricts `implements`).

MiniPiku commented:
@eernstg just a follow-up comment that the said changes have been made 
kindly take a look

eernstg reviewed (commented):
Thanks! It looks good to me now.

I haven't pressed the 'Approve' button because I'm from the language team and the final acceptance should be provided by the site-www team.

MiniPiku commented on src/content/effective-dart/_toc.md:
```diff
@@ -199,8 +199,7 @@ the project:
 * <a href='/effective-dart/design#avoid-defining-a-class-that-contains-only-static-members'>AVOID defining a class that contains only static members.</a>
-* <a href='/effective-dart/design#avoid-extending-a-class-that-isnt-intended-to-be-subclassed'>AVOID extending a class that isn't intended to be subclassed.</a>
-* <a href='/effective-dart/design#do-document-if-your-class-supports-being-extended'>DO document if your class supports being extended.</a>
+* <a href='/effective-dart/design#do-use-class-modifiers-to-control-if-your-class-can-be-extended'>DO use class modifiers to control if your class can be extended</a>
```
fixed :)

MiniPiku commented on src/content/effective-dart/_toc.md:
```diff
@@ -199,8 +199,7 @@ the project:
 
 * <a href='/effective-dart/design#avoid-defining-a-one-member-abstract-class-when-a-simple-function-will-do'>AVOID defining a one-member abstract class when a simple function will do.</a>
 * <a href='/effective-dart/design#avoid-defining-a-class-that-contains-only-static-members'>AVOID defining a class that contains only static members.</a>
-* <a href='/effective-dart/design#avoid-extending-a-class-that-isnt-intended-to-be-subclassed'>AVOID extending a class that isn't intended to be subclassed.</a>
```
yep my bad i thought that this section was removed from the `design.md` file that's why i remove the link too
lemme readd that 
thanks for pointing it out

MiniPiku replied:
done

johnpryan commented:
Can you revert https://github.com/dart-lang/site-www/pull/6449/commits/4dc729d50c43fbec9638efb1dc89dbe08073f674, thanks!

MiniPiku commented:
@johnpryan done 
but before this this it was giving an Github PR run issue with the anchor link idk why

johnpryan commented on src/content/language/patterns.md:
```diff
@@ -360,14 +360,14 @@ double calculateArea(Shape shape) => switch (shape) {
 
 [Map][] and [list][] patterns work well for destructuring key-value pairs in
-JSON data:
+deserialized data, such as data parsed from JSON:
```
Revert changes to this file, please!

johnpryan commented on examples/language/lib/patterns/json.dart:
```diff
@@ -2,18 +2,18 @@
 void main() {
   // #docregion json-1
-  var json = {
+  var data = {
```
Revert changes to this file, please!

MiniPiku replied:
same with this one 
https://github.com/dart-lang/site-www/pull/6449#discussion_r1980158012

johnpryan replied:
I merged this branch with `main` and this seemed to resolve the issue.

johnpryan reviewed (approved):
LGTM

antfitch reviewed (approved):
Looks great!

antfitch commented:
/gcbrun

antfitch commented:
@MiniPiku it looks like a link is broken. Can you fix that?

antfitch reviewed (changes requested):
One last change to fix a broken link. Looks great!

MiniPiku commented:
@antfitch kindly take a look 
i guess this should work

antfitch commented:
/gcbrun

antfitch reviewed (approved):
Looks great!
